pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,237 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package ssa
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/schemamutation"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+// cachingUnstructuredExtractor is a caching implementation of v1.UnstructuredExtractor
+// using OpenAPI V3 discovery information.
+// TODO(erhan): try to upstream this code in kubernetes
+type cachingUnstructuredExtractor struct {
+	// added as field to not break the interface for other funcs, instantiated at each reconcile
+	ctx   context.Context
+	cache *GVKParserCache
+	dc    discovery.DiscoveryInterface
+}
+
+// NewCachingUnstructuredExtractor returns a new cachingUnstructuredExtractor
+func NewCachingUnstructuredExtractor(ctx context.Context, dc discovery.DiscoveryInterface, cache *GVKParserCache) (applymetav1.UnstructuredExtractor, error) {
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+		ctx:   ctx,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+func discoveryPaths(ctx context.Context, rc rest.Interface) (map[string]OpenAPIGroupVersion, error) {
+	data, err := rc.Get().
+		AbsPath("/openapi/v3").
+		Do(ctx).
+		Raw()
+
```

Comment:
```suggestion
	data, err := rc.Get().AbsPath("/openapi/v3").Do(ctx).Raw()
```

No empty lines before error handling.